### PR TITLE
Only handle broadcasted messages meant to be received

### PR DESCRIPTION
**Pull request by andeplane**
## References

Attempt to solve https://github.com/jupyterlite/jupyterlite/issues/1074 (also https://github.com/jupyterlite/jupyterlite/issues/1126 and https://github.com/jupyterlite/jupyterlite/issues/1089)

## Code changes

Ref https://github.com/jupyterlite/jupyterlite/issues/1074, sometimes `_onMessage` in [broadcast.ts](https://github.com/jupyterlite/jupyterlite/blob/main/packages/contents/src/broadcast.ts#L61) gets a message that does not meet the expected structure of the function. This PR makes sure these messages are ignored by this receiever.

### CODE REVIEW

**Comment by andeplane** [comment]
Tested the PR server in our environment now, and it seems to work, as a message is being ignored with this change. However, a better solution would be to figure out why the message is sent there in the first place.

image

**Comment by martinRenou** [comment]
Thanks a lot for opening a PR, this was on my TODO list for weeks!

> However, a better solution would be to figure out why the message is sent there in the first place

I agree. Were you able to see which message was sent in the first place? This request should be built and sent in https://github.com/jupyterlite/jupyterlite/blob/2ee9cba529cb514c00130924c355d8696fd4562c/packages/contents/src/drivefs.ts#L301, but I don't see a reason for the content to be `null`. We may want to figure out what the message originated from.
We may also want to add there the check for data==null before sending the message. If we do the check in the service worker, I fear the kernel may fail and get stuck later on.

**Comment by andeplane** [comment]
Yeah! I have some insight:

If you look at this screenshot, you see that the `e.data` is an object that matches the `response` being created [here](https://github.com/jupyterlite/jupyterlite/blob/main/packages/contents/src/broadcast.ts#L114), which seems to be sent later in the same `_onMessage` function [here](https://github.com/jupyterlite/jupyterlite/blob/main/packages/contents/src/broadcast.ts#L157).

I don't understand why it doesn't always fail after reading it, but I am definitely not understanding this properly yet.

image

**Comment by martinRenou** [comment]
Ok 😕  Would you be able to put a `console.log("message", JSON.stringify(data))` in here: https://github.com/jupyterlite/jupyterlite/blob/2ee9cba529cb514c00130924c355d8696fd4562c/packages/contents/src/drivefs.ts#L304 ?

I would test it myself, I am not able to reproduce this unfortunately, I've only seen it on other people's env.

**Comment by dlcognite** [comment]
When debugging this, what we noticed is that
- we always (no matter in iframe, or as standalone) get an event where the request.method is  `getattr` in the `_onMessage` of the broadcast.ts file.
- at the end of the file it will `_postMessage` to the `_channel` this body
```
{
        dev: 1,
        nlink: 1,
        uid: 0,
        gid: 0,
        rdev: 0,
        size: model.size || 0,
        blksize: BLOCK_SIZE,
        blocks: Math.ceil(model.size || 0 / BLOCK_SIZE),
        atime: model.last_modified, // TODO Get the proper atime?
        mtime: model.last_modified,
        ctime: model.created,
        timestamp: 0,
      }
      ```
      
      notice here that the body here doesn't have a `method`
      
     when in iframe, however the same channel would receive this new message missing "method", and fail. but when running this not in iframe, it works fine... so somehow in the iframe, the `_postMessage` comes back to the same broadcast channel...

**Comment by andeplane** [comment]
Ok, we think we know what it is. When two or more tabs are open, messages sent from `broadcast.ts` are being received in the other tab's `BroadcastChannelWrapper._onMessage`, which does not have the expected format. Normally you have one sender and one receiver on the `/api/drive.v1` channel, but it gets mixed up with multiple tabs.

The right solution (I think) is to have a [session id](https://github.com/jupyterlite/jupyterlite/blob/main/packages/kernel/src/kernel.ts#L18) that is being attached to all messages so the other receivers can ignore messages.

But on the other hand, since the file system is shared across all tabs it actually does make sense to let tabs communicate, in which case a slight modification this PR might be the right solution.

What do you think @martinRenou?

File: packages/contents/src/drivefs.ts
Comment (martinRenou) [inline packages/contents/src/drivefs.ts:360]: Could this be a `const SENDER = 'drivefs.ts'` at the top of the file?

**Review by martinRenou — COMMENTED** [review]
Thanks for continuing exploring this!

It's unclear to me where the other message comes from.

**Comment by andeplane** [comment]
> Thanks for continuing exploring this!
> 
> It's unclear to me where the other message comes from.

So if you have two tabs A and B, both running the notebook.
`A.broadcast.ts` is sending a message on the `/api/drive.v1` channel that is meant to be received by `A.drivefs.ts`. `A.drivefs.ts` receives it and all is good. But here is the thing, since this channel is shared across tabs, `B.drivefs.ts` receives it (which is ok), but `B.broadcast.ts` also receives it, and that's why it breaks because `B.broadcast.ts` expects messages with a different payload.

File: packages/contents/src/drivefs.ts
Comment (andeplane) [reply packages/contents/src/drivefs.ts]: Fixed that now

**Review by martinRenou — APPROVED** [review]
LGTM! @jtpio if you're fine with this PR, it would be great to make a patch release with it

**Comment by jtpio** [comment]
Thanks all for looking into this!

Looks like the UI tests failure might be relevant though. 

Otherwise yes it would be great to get this fix released after.

**Comment by andeplane** [comment]
Ok great, I’ll fix the ui tests.

**Comment by jtpio** [comment]
Looking at the report, the UI tests seem to be stuck while executing the cell: 

image

**Comment by jtpio** [comment]
So trying with the ReadTheDocs preview for this PR (https://jupyterlite--1075.org.readthedocs.build/en/1075/_static/lab/index.html) in a fresh new browser tab on Chromium gives the following error:

image

**Comment by andeplane** [comment]
Yep, I was testing a thing that is in the latest commit. Will do some work tomorrow and fix UI tests and make sure it works.

**Comment by andeplane** [comment]
Seems to be working, please verify :)

**Comment by jtpio** [comment]
Thanks @andeplane for your work on this!

Looks like the ReadTheDocs preview is still giving the same error? (trying in a Chromium private window)

https://jupyterlite--1075.org.readthedocs.build/en/1075/_static/lab/index.html

**Comment by andeplane** [comment]
> 

Yep I see that, will be sure to test with incognito. It seems like it breaks when we have the
`if (event.data?.sender !== 'broadcast.ts')` test, which means some messages are being sent elsewhere. Testing without it now, and will find out who that sender is.

**Comment by andeplane** [comment]
@martinRenou back working on this now, seems like I need your approval to run workflows!

**Comment by andeplane** [comment]
@jtpio I think I need some help here. I know for sure that UI tests start failing with the test on `event.data?.sender` which could make sense if the tests somehow communicate with the file system using the broadcast channel and doesn't set the sender. However the UI tests seems to use functions provided by JupyterLab?
```
const name = await page.notebook.createNew();

    await page.notebook.setCell(0, 'markdown', '## This is a markdown cell');
    await page.notebook.addCell('raw', 'This is a raw cell');
    await page.notebook.addCell('code', '2 + 2');

    await page.notebook.run();
    await page.notebook.save();
```

(I can't find e.g. `getCellTextOutput` anywhere in the source code for JupyterLite. 

Question is then why this change does anything to the tests (I can manually open notebooks and run them, so doesn't happen on _my_ browser, hehe).

**Comment by jtpio** [comment]
@andeplane not sure it was only the tests failing. According to the comment above (https://github.com/jupyterlite/jupyterlite/pull/1075#issuecomment-1549085438) the failure seemed relevant and also happening on the ReadTheDocs demo.

Just kicked the CI, let's see if it passes.

**Comment by andeplane** [comment]
Right, forgot it after all this time, sorry. I see that it is true on readthedocs. I can't reproduce locally (works when I build and run using gitpod), so the only reproducible environment I have is building here, so expect some noise while I debug through readthedocs 😂

But that is great news tbh, because then it isn't something weird like I was trying to claim above. I'm still confident the underlying issue is the messaging like explained earlier, so I'll keep digging. Thanks!

**Comment by andeplane** [comment]
I figured out why it didn't work earlier (me being stupid was one good reason!), and changed the logic so we mark the messages being sent using serviceWorker to broadcast.ts with `receiver: 'broadcast.ts'`, so all messages being sent to itself will be ignored. 

This now works locally, and I can open multiple tabs with [ReadTheDocs](https://jupyterlite--1075.org.readthedocs.build/en/1075/_static/lab/index.html), and the tests pass. 

Would love for you folks to give it a go as well.

**Review by jtpio — APPROVED** [review]
Thanks!

**Comment by jtpio** [comment]
Planning to make a patch release with this change on Monday 21.

**Comment by ogrisel** [comment]
Feel free to ping me once this is done so that I can trigger a CI update for the environment used to reproduce https://github.com/jupyterlite/jupyterlite/issues/1126 so that we can confirm that this can close it.

**Comment by jtpio** [comment]
@ogrisel `0.1.2` has been released: https://github.com/jupyterlite/jupyterlite/releases/tag/v0.1.2

**Pull request merged**
